File: machine-learning/ex1/featureNormalize.m
function [X_norm, mu, sigma] = featureNormalize(X)
  X_norm = X;
  mu = zeros(1, size(X, 2));
  sigma = zeros(1, size(X, 2));
  
  num_features = size(X,2);
  
  for x=1:num_features,
    mu(x) = mean(X(:,x));
    sigma(x) = std(X(:,x));
    
    X_norm(:,x) = (X_norm(:,x)-mu(x)) / sigma(x);
  end
  
end

File: machine-learning/ex1/python/featureNormalize.py
import numpy as np

def featureNormalize(X):
    X_norm = X.copy()
    mu = np.zeros((1, X.shape[1]))
    sigma = np.zeros((1, X.shape[1]))

    for i in range(X_norm.shape[1]):
        mu[:,i] = np.mean(X[:,i])
        sigma[:,i] = np.std(X[:,i])
        X_norm[:,i] = (X[:,i] - mu[:,i]) / sigma[:,i]

    return X_norm, mu, sigma